### PR TITLE
Rewrote /federation_resolvers - POST

### LINKED ISSUE #3796: Rewrite /federation_resolvers to Go
## I'm submitting an ...
improvement request (usability, performance, tech debt, etc.)

## Traffic Control components affected ...
- Documentation (possibly)
- Traffic Ops

## Current behavior:
`/federation_resolvers` is handled by Perl

## Expected / new behavior:
`/federation_resolvers` should be handled by Go.

### Minimum API Version
1.1

### Request Methods
- [ ] `GET`
- [ ] `POST`

## Minimal reproduction of the problem with instructions:
N/A

### PR DESCRIPTION
**Pull request by ocket8888**
## What does this PR (Pull Request) do?
- [x] This PR fixes #3796

Rewrite `/federation_resolvers` to Go. Also adds Traffic Ops and Traffic Control client code and associated tests (for the TO client, not the Python).

API version 1.3 makes no (intentional) changes from the Perl handler's behavior, but the 1.4 GET handler adds the common `last_updated` field to the response.

Finally, although this rewrite does not include `/federations/{{ID}}`, it does include one client method (for deleting Federation Resolvers) that uses that path.

## Which Traffic Control components are affected by this PR?
- Documentation
- Traffic Control Client (Python)
- Traffic Ops Client (Golang)
- Traffic Ops

## What is the best way to verify this PR?
Run API and unit tests. Otherwise, just do GET/POST/DELETE requests to `/federation_resolvers` as documented.

## The following criteria are ALL met by this PR
- [x] This PR includes tests
- [x] This PR includes documentation
- [x] An update to CHANGELOG.md is not necessary
- [x] This PR includes any and all required license headers
- [x] This PR does not include a database migration
- [x] This PR **DOES NOT FIX A SERIOUS SECURITY VULNERABILITY**

### CODE REVIEW

File: traffic_ops/client/federation_resolver.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+import "encoding/json"
+import "fmt"
+import "net/http"
+import "net/url"
+import "strconv"
```
Comment (awilliams) [inline traffic_ops/client/federation_resolver.go]: All files in the `client` package use the following import style. Consider matching for consistency.

```go
import (
  "a"
  "b"
  ...
)
```

Comment (ocket8888) [reply traffic_ops/client/federation_resolver.go]: That's not true. invalidationjobs and traffic_stats use the same style - as well as several others outside of the `client` package. and we don't have a style guide - only `go fmt` is enforced (which doesn't care).

File: docs/source/api/federation_resolvers.rst [hunk 2]
```diff
@@ -19,6 +19,87 @@
...
+	| Name       | Required | Description                                                                                         |
+	+============+==========+=====================================================================================================+
+	| id         | no       | Return only the Federation Resolver identified by this integral, unique identifier                  |
+	+------------+----------+-----------------------------------------------------------------------------------------------------+
+	| ipAddress  | no       | Return only the Federation Resolver(s) that has/have this IP Address                                |
+	+------------+----------+-----------------------------------------------------------------------------------------------------+
+	| type       | no       | Return only the Federation Resolvers of this :term:`Type`                                           |
+	+------------+----------+-----------------------------------------------------------------------------------------------------+
+	| orderby    | no       | Choose the ordering of the results - must be the name of one of the fields of the objects in the    |
```
Comment (mhoppa) [inline docs/source/api/federation_resolvers.rst]: I think as we add new documentation we should fix `Choose the ordering of the results - must be the name of one of the fields of the objects in the ``response`` array` as that is not correct

Comment (mhoppa) [reply docs/source/api/federation_resolvers.rst]: We cant do it by lastUpdated

File: docs/source/api/federation_resolvers.rst [hunk 3]
```diff
@@ -30,65 +111,57 @@ Creates a new federation resolver.
...
 -----------------
 :ipAddress: The IP address of the resolver - may be IPv4 or IPv6
-:typeId:    The integral, unique identifier of the type of resolver being created - will *represent* one of:
+:typeId:    The integral, unique identifier of the :term:`Type` of resolver being created
 
-	RESOLVE4
-		Resolver is for IPv4 addresses and ``ipAddress`` is IPv4
-	RESOLVE6
-		Resolver is for IPv6 addresses and ``ipAddress`` is IPv6
+	.. caution:: This field should only ever be an identifier for one of the :term:`Types` "RESOLVE4" or "RESOLVE6", but there is **no protection for this built into Traffic Ops** and therefore **any valid** :term:`Type` **identifier will be silently accepted by Traffic Ops** and so care should be taken to ensure that these :term:`Types` are properly identified. If any :term:`Type` besides "RESOLVE4" or "RESOLVE6" is identified, the resulting resolver *will* **not** *work*.
```
Comment (mhoppa) [inline docs/source/api/federation_resolvers.rst]: should we fix that in the rewrite? I couldnt imagine someone relying on this endpoint accepting invalid typeIds

File: traffic_ops/traffic_ops_golang/federation_resolvers/federation_resolvers.go [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+          ) AS type
+`
+
+const readQuery = `
+SELECT federation_resolver.id,
+       federation_resolver.ip_address,
+       federation_resolver.last_updated,
+       type.name AS type
+FROM federation_resolver
+LEFT OUTER JOIN type ON type.id = federation_resolver.type
```
Comment (mhoppa) [inline traffic_ops/traffic_ops_golang/federation_resolvers/federation_resolvers.go]: reason for LEFT OUTER here compared to just LEFT JOIN?

File: docs/source/api/federation_resolvers.rst (hunk 2 above)
Comment (ocket8888) [reply docs/source/api/federation_resolvers.rst]: My plan was to add a footnote clarifying that the `lastUpdated` field is unavailable for that. Do you think a different wording right in the table would be better?

Comment (mhoppa) [reply docs/source/api/federation_resolvers.rst]: that would be fine

File: docs/source/api/federation_resolvers.rst (hunk 3 above)
Comment (ocket8888) [reply docs/source/api/federation_resolvers.rst]: :man_shrugging: I sort of wanted to, but it's been said that we should:

> _"... lean towards resisting the temptation during the perl -> go rewrites to change how things work as much as possible; even if it's wrong.  That way if something goes wrong unexpectedly, it's easier to determine which was the cause."_

Restricting the allowed values of `typeId` would be a breaking change so...

File: traffic_ops/traffic_ops_golang/federation_resolvers/federation_resolvers.go (hunk 4 above)
Comment (ocket8888) [reply traffic_ops/traffic_ops_golang/federation_resolvers/federation_resolvers.go]: They're the same; `OUTER` is optional, just more explicit.

**Comment by ocket8888** [comment]
Perl let you do the same thing. It's probably as a direct consequence of never validating the "Type". 

Honestly, that should never come into the equation, the payload should just be an IP Address of arbitrary version. Any special handling of it based on the protocol version can be figured out entirely on the back-end, and that's what I'd push for if this was being designed today.

File: docs/source/api/federation_resolvers.rst (hunk 2 above)
Comment (ocket8888) [reply docs/source/api/federation_resolvers.rst]: I'm working on a branch to do this globally, since the problem exists globally.

Comment (mhoppa) [reply docs/source/api/federation_resolvers.rst]: awesome :) above and beyond 🚀

**Pull request merged**